server/src/services/documentSymbol/onDocumentSymbol.ts

```diff
@@ -109,11 +106,11 @@ export function onDocumentSymbol(serverState: ServerState) {
         // } while (kursor.goToNext());
 
         span = transaction.startChild({ op: "walk-generate-symbols" });
-        while ((node = cursor.findRuleWithKind(ruleKinds)) !== null) {
+        while (cursor.goToNextRuleWithKinds(ruleKinds)) {
+          const node = cursor.node() as RuleNode;
```

Comment:
note to self: add a task to expose `node.asRule()` and `node.asToken()` to the TS API.

Reply:
if I understand correctly, this would probably work as `-> Option<T>` or `cst.XYZNode | null`, which wouldn't give us much here (would force us to use the non-null assertion). I can check for something like `"text" in node` to appease TS in a different way, but I didn't want to add further noise here.

Ideally, since the `goToNext` specifies that it has to go to a Rule, it would, as before, return an `Option<RuleNode>` but would be clearer that now that it can't be stuck at the same position (when called multiple times) since the name explicitly refers to a "next" node rather than just "finding" one.

Reply:
I was going to suggest using [assertion functions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions):

```ts
function assertToken(arg: any) asserts arg is TokenNode {}
```

And [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates):

```ts
function isToken(): arg is TokenNode {}
```

This is used by TS type narrowing, and it idiomatically reports errors through `AssertionError` if needed.
I imagine this is going to be a common use case for CST users?

Reply:
Yeah, we could have more idiomatic checks like type predicates callable directly on the node (if it's possible; are type guards/predicates callable with `this`?) rather than hacky `'xyz' in node` checks, agreed there.

I was referring specifically to the proposed `node.asXYZ()` above and assumed it would have the `-> Option<T> (Rust)` or `: cst.XYZNode | null` (exposed TS) result type, which is where I was not sure if it would improve the status quo, since the caller would have to perform another check, e.g.
```ts
// ...snip make sure the node points to a rule node
let actuallyRule;
const result = node.asRule();
if (result == null) { continue/break/return/throw (diverge); }
else { actuallyRule = result; }
```

or
```ts
// ...snip, make sure the node points to a rule node
let actuallyRule = node.asRule()!; // safe in practice but annoying since we have to silence ESLint
```